In [391]:

import os

from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

import pandas as pd
import numpy as np

In [392]:
pd.set_option('display.max_columns', None)
df = pd.read_csv(os.path.join("..","in","fifa.csv"))

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club                                     Club Logo    Value  \
0         FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png  €110.5M   
1             Juventus   https://cdn.sofifa.org/teams/2/light/45.png     €77M   
2  Paris Saint-Germain   https://cdn.sofifa.org/teams/2/light/73.png  €118.5M   
3    Manchester United   https://cdn.sofifa.org/teams/2/light/11.png     €72M   
4      Manchester City   https://cdn.sofifa.org/teams/2/light/10.png    €102M   

    Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  €565K     2202           Left                       5.0        4.0   
1  €405K     2228          Right                       5.0        4.0   
2  €290K     2143          Right                       5.0        5.0   
3  €260K     1471          Right                       4.0        3.0   
4  €355K     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Real Face Position  Jersey Number  \
0          4.0  Medium/ Medium       Messi       Yes       RF           10.0   
1          5.0       High/ Low  C. Ronaldo       Yes       ST            7.0   
2          5.0    High/ Medium      Neymar       Yes       LW           10.0   
3          1.0  Medium/ Medium        Lean       Yes       GK            1.0   
4          4.0      High/ High      Normal       Yes      RCM            7.0   

         Joined Loaned From Contract Valid Until Height  Weight    LS    ST  \
0   Jul 1, 2004         NaN                 2021    5'7  159lbs  88+2  88+2   
1  Jul 10, 2018         NaN                 2022    6'2  183lbs  91+3  91+3   
2   Aug 3, 2017         NaN                 2022    5'9  150lbs  84+3  84+3   
3   Jul 1, 2011         NaN                 2020    6'4  168lbs   NaN   NaN   
4  Aug 30, 2015         NaN                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     8

# Cleaning Data

# Eliminación de las columnas no relevantes

Ya que nuestro objetivo principal hacer predicciones sobre los valores de los jugadores, miramos qué caracteríaticas realmente están intrísicamente relacionadas con la columna "Value". Por ejemplo, una columna que especifica una habilidad del jugador será importante para determinar el valor final, en cambio, la columna "Photo" no puede ser relacionada de ninguna forma para determinar ese valor del jugador. 

Estudiamos columnas por columnas del df y eliminamos las menos relevantes:

No nos interesan las primeras dos columnas, así que las eliminamos:

In [395]:
df = df.iloc[:, 2:]
df.head()

Name  Age  \
0                L. Messi   31   
1       Cristiano Ronaldo   33   
2               Neymar Jr   26   
3                  De Gea   27   
4            K. De Bruyne   27   
...                   ...  ...   
18202        J. Lundstram   19   
18203  N. Christoffersson   19   
18204           B. Worman   16   
18205      D. Walker-Rice   17   
18206           G. Nugent   16   

                                                Photo Nationality  \
0      https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1       https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2      https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3      https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4      https://cdn.sofifa.org/players/4/19/192985.png     Belgium   
...                                               ...         ...   
18202  https://cdn.sofifa.org/players/4/19/238813.png     England   
18203  https://cdn.sofifa.org/players/4/19/243165.png      Sweden   
18204  https://cdn.sofifa.org/players/4/19/241638.png     England   
18205  https://cdn.sofifa.org/players/4/19/246268.png     England   
18206  https://cdn.sofifa.org/players/4/19/246269.png     England   

                                      Flag  Overall  Potential  \
0      https://cdn.sofifa.org/flags/52.png       94         94   
1      https://cdn.sofifa.org/flags/38.png       94         94   
2      https://cdn.sofifa.org/flags/54.png       92         93   
3      https://cdn.sofifa.org/flags/45.png       91         93   
4       https://cdn.sofifa.org/flags/7.png       91         92   
...                                    ...      ...        ...   
18202  https://cdn.sofifa.org/flags/14.png       47         65   
18203  https://cdn.sofifa.org/flags/46.png       47         63   
18204  https://cdn.sofifa.org/flags/14.png       47         67   
18205  https://cdn.sofifa.org/flags/14.png       47         66   
18206  https://cdn.sofifa.org/flags/14.png       46         66   

                      Club                                       Club Logo  \
0             FC Barcelona    https://cdn.sofifa.org/teams/2/light/241.png   
1                 Juventus     https://cdn.sofifa.org/teams/2/light/45.png   
2      Paris Saint-Germain     https://cdn.sofifa.org/teams/2/light/73.png   
3        Manchester United     https://cdn.sofifa.org/teams/2/light/11.png   
4          Manchester City     https://cdn.sofifa.org/teams/2/light/10.png   
...                    ...                                             ...   
18202      Crewe Alexandra    https://cdn.sofifa.org/teams/2/light/121.png   
18203       Trelleborgs FF    https://cdn.sofifa.org/teams/2/light/703.png   
18204     Cambridge United   https://cdn.sofifa.org/teams/2/light/1944.png   
18205      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   
18206      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   

         Value   Wage  Special Preferred Foot  International Reputation  \
0      €110.5M  €565K     2202           Left                       5.0   
1         €77M  €405K     2228          Right                       5.0   
2      €118.5M  €290K     2143          Right                       5.0   
3         €72M  €260K     1471          Right                       4.0   
4        €102M  €355K     2281          Right                       4.0   
...        ...    ...      ...            ...                       ...   
18202     €60K    €1K     1307          Right                       1.0   
18203     €60K    €1K     1098          Right                       1.0   
18204     €60K    €1K     1189          Right                       1.0   
18205     €60K    €1K     1228          Right                       1.0   
18206     €60K    €1K     1321          Right                       1.0   

       Weak Foot  Skill Moves       Work Rate   Body Type Real Face Position  \
0            4.0          4.0  Medium/ Medium       Messi       Yes     

Analizando los datos, podemos observar que también se pueden eliminar las siguientes columnas del dataframe ya que no aportan un significado para calcular el valor de un jugador:

Las habilidades de un jugador son las que determinan su precio real. En cambuio, las columnas como nombre, nacionalidad, foto, bandera, logo del club y cara no determinan de ninguna forma este valor, por lo tanto, no nos interesan y las podemos eliminar.

Altura: un jugador puede tener buenas habilidades independientemente de su altura. Por ejemplo, C. Ronaldo es uno de los más altos y Messi no es alto. Sin embargo, los dos se encuentran en posiciones altas. Teniendo en cuenta lo anterior, se puede eliminar esta columna.

Peso: podemos hacer la misma suposición para el peso que en la altura. También lo eliminamos.

Pie preferido: interesa que los jugadores diestros se encuentren en la posición derecha y los zurdos en la posición izquierda. Pero estos jugadores tienen el pie débil bastante entrenado, no importa mucho el pie preferido. Si tenemos dos jugadores con mismas habilidades y condiciones, pero uno es diestro y el otro es zurdo, y además, uno se encuentra en la posición delantera derecha y el otro en la izquierda, los valores de estos jugadores serán los mismos.

Salario: no es el salario que determina el valor del jugador, sino es el jugador que condiciona el salario.

Contato válido hasta: solamente es un contrato, eso sí, puede ser que los precios en los contratos de un año sean diferente de los contratos de otros años, pero esto afectaría a todos los jugadores por igual. Así que, lo podemos eliminar.

Tipo de cuerpo: es indiferente para ser un jugador bueno.

Desde la columna "LS" hasta "RB": no sabemos el significado de esas columnas, tampoco aparece de forma clara en otros recursos, podemos suponer que no son columnas importante.


In [396]:
col_to_drop=["Name","Nationality","Photo","Flag","Club Logo","Real Face","Height","Weight","Preferred Foot","Wage","Jersey Number","Joined","Contract Valid Until","Body Type","LS","ST","RS","LW","LF","CF","RF","RW","LAM","CAM","RAM","LM","LCM","CM","RCM","RM","LWB","LDM","CDM","RDM","RWB","LB","LCB","CB","RCB","RB"]
df=df.drop(axis=1,columns=col_to_drop)

# Handling NaN's

Imprimimos una lista de aquellas columnas que contengan almenos un valor NaN

In [397]:
df.columns[df.isna().any()].tolist()

['Club',
 'International Reputation',
 'Weak Foot',
 'Skill Moves',
 'Work Rate',
 'Position',
 'Loaned From',
 'Crossing',
 'Finishing',
 'HeadingAccuracy',
 'ShortPassing',
 'Volleys',
 'Dribbling',
 'Curve',
 'FKAccuracy',
 'LongPassing',
 'BallControl',
 'Acceleration',
 'SprintSpeed',
 'Agility',
 'Reactions',
 'Balance',
 'ShotPower',
 'Jumping',
 'Stamina',
 'Strength',
 'LongShots',
 'Aggression',
 'Interceptions',
 'Positioning',
 'Vision',
 'Penalties',
 'Composure',
 'Marking',
 'StandingTackle',
 'SlidingTackle',
 'GKDiving',
 'GKHandling',
 'GKKicking',
 'GKPositioning',
 'GKReflexes',
 'Release Clause']

A continuación eliminaremos aquellas columnas que tengan la mayoría de los valores NaN. Con mayoría podemos entender aquellas que tengan más de 1/4 de valores NaN (es una suposición personal). No podemos asignar unos en estas columnas ya que faltan bastante, lo más lógico sería eliminarlas. Para ver el número de filas y columnas que tiene nuestro df:

In [399]:
df.shape

(18207, 47)

Nuestro data frame tiene 18207/4 = 4552 filas aprox. Entonces nuestro threshold (número de valores NaN mínimo que tiene que tener una columna para que sea elminada del df) será 4552 -> ~=4550

In [400]:
df = df.dropna(thresh=4550,axis=1)

Age  Overall  Potential                 Club    Value  Special  \
0       31       94         94         FC Barcelona  €110.5M     2202   
1       33       94         94             Juventus     €77M     2228   
2       26       92         93  Paris Saint-Germain  €118.5M     2143   
3       27       91         93    Manchester United     €72M     1471   
4       27       91         92      Manchester City    €102M     2281   
...    ...      ...        ...                  ...      ...      ...   
18202   19       47         65      Crewe Alexandra     €60K     1307   
18203   19       47         63       Trelleborgs FF     €60K     1098   
18204   16       47         67     Cambridge United     €60K     1189   
18205   17       47         66      Tranmere Rovers     €60K     1228   
18206   16       46         66      Tranmere Rovers     €60K     1321   

       International Reputation  Weak Foot  Skill Moves       Work Rate  \
0                           5.0        4.0          4.0  Medium/ Medium   
1                           5.0        4.0          5.0       High/ Low   
2                           5.0        5.0          5.0    High/ Medium   
3                           4.0        3.0          1.0  Medium/ Medium   
4                           4.0        5.0          4.0      High/ High   
...                         ...        ...          ...             ...   
18202                       1.0        2.0          2.0  Medium/ Medium   
18203                       1.0        2.0          2.0  Medium/ Medium   
18204                       1.0        3.0          2.0  Medium/ Medium   
18205                       1.0        3.0          2.0  Medium/ Medium   
18206                       1.0        3.0          2.0  Medium/ Medium   

      Position  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0           RF      84.0       95.0             70.0          90.0     86.0   
1           ST      84.0       94.0             89.0          81.0     87.0   
2           LW      79.0       87.0             62.0          84.0     84.0   
3           GK      17.0       13.0             21.0          50.0     13.0   
4          RCM      93.0       82.0             55.0          92.0     82.0   
...        ...       ...        ...              ...           ...      ...   
18202       CM      34.0       38.0             40.0          49.0     25.0   
18203       ST      23.0       52.0             52.0          43.0     36.0   
18204       ST      25.0       40.0             46.0          38.0     38.0   
18205       RW      44.0       50.0             39.0          42.0     40.0   
18206       CM      41.0       34.0             46.0          48.0     30.0   

       Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0           97.0   93.0        94.0         87.0         96.0          91.0   
1           88.0   81.0        76.0         77.0         94.0          89.0   
2           96.0   88.0        87.0         78.0         95.0          94.0   
3           18.0   21.0        19.0         51.0         42.0          57.0   
4           86.0   85.0        83.0         91.0         91.0          78.0   
...          ...    ...         ...          ...          ...           ...   
18202       42.0   30.0        34.0         45.0         43.0          54.0   
18203       39.0   32.0        20.0         25.0         40.0          41.0   
18204       45.0   38.0        27.0         28.0         44.0          70.0   
18205       51.0   34.0        32.0         32.0         52.0          61.0   
18206       43.0   40.0        34.0         44.0         51.0          57.0   

       SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0             86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1             91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2             90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3          

Ya que los datos son de la FIFA, eliminamos aquellos jugadores cuyos clubs no pertenecen en FIFA:

In [401]:
df = df.dropna(subset = ['Club'])

Miramos el número de los valores NA's que tiene cada columna:

In [402]:
(df.isnull().sum()).tolist()

[0,
 0,
 0,
 0,
 0,
 0,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 48,
 1323]

Se vé que hay 48 filas que tienen valores nans para muchas columnas. Entonces, no intersará eliminar dichas filas además de que no son muchas comparadas con el número de filas totales. Bastaría si trabajamos con una única columna para eliminar las filas.

Véamos de qué filas se tratan:

In [403]:
  for col in df:
    print(col)
    print(df[col].isnull().sum())

Age
0
Overall
0
Potential
0
Club
0
Value
0
Special
0
International Reputation
48
Weak Foot
48
Skill Moves
48
Work Rate
48
Position
48
Crossing
48
Finishing
48
HeadingAccuracy
48
ShortPassing
48
Volleys
48
Dribbling
48
Curve
48
FKAccuracy
48
LongPassing
48
BallControl
48
Acceleration
48
SprintSpeed
48
Agility
48
Reactions
48
Balance
48
ShotPower
48
Jumping
48
Stamina
48
Strength
48
LongShots
48
Aggression
48
Interceptions
48
Positioning
48
Vision
48
Penalties
48
Composure
48
Marking
48
StandingTackle
48
SlidingTackle
48
GKDiving
48
GKHandling
48
GKKicking
48
GKPositioning
48
GKReflexes
48
Release Clause
1323


Como ya hemos dicho, bastaría eliminar estas filas haciendo referencia a una de las varias columnas que las contien. Por ejemplo, la columna Posición:

In [404]:
df = df.dropna(axis=0, how='any',subset=["Position"])

In [405]:
  for col in df:
    print(col)
    print(df[col].isnull().sum())

Age
0
Overall
0
Potential
0
Club
0
Value
0
Special
0
International Reputation
0
Weak Foot
0
Skill Moves
0
Work Rate
0
Position
0
Crossing
0
Finishing
0
HeadingAccuracy
0
ShortPassing
0
Volleys
0
Dribbling
0
Curve
0
FKAccuracy
0
LongPassing
0
BallControl
0
Acceleration
0
SprintSpeed
0
Agility
0
Reactions
0
Balance
0
ShotPower
0
Jumping
0
Stamina
0
Strength
0
LongShots
0
Aggression
0
Interceptions
0
Positioning
0
Vision
0
Penalties
0
Composure
0
Marking
0
StandingTackle
0
SlidingTackle
0
GKDiving
0
GKHandling
0
GKKicking
0
GKPositioning
0
GKReflexes
0
Release Clause
1275


Como podéis ver, la columna "Release Clause" tiene 1275 valores nans. Tenemos dos opciones para arreglar esto:

1) Eliminar las columna si sus significados no es relevante para nuestro objetivo

2) Dar ciertos valores a esos valores nans para columnas que podrían conservar información importante.

La columna "Release Clause" deberíamos de conservar ya que aporta información importante. Así que vamos arreglar la columna "Release Clause" sustituiendo valores nans por otros valores.


Tenemos que elegir de varias deciciones que podemos tomar:

1) Dar un valor concreto a todos los nans de la columna (no nos interesa)

2) Asignar el mismo valor que la fila de arriba, o bien la fila de abajo (backward or forward)

3) Hacer una interpolación

4) etc

Está claro que la mejor opción es la tercera, así que elegiremos esta opción.

## Tratar los NaNs

Convertimos los valores de coste económico en float, sin letras, con la ayuda de la siguente función:

In [406]:
def value_to_float(x):
    """
    From K and M to float.
    
    """
    if 'X' in x:
      return 0.0
    x = x.replace('€', '')
    ret_val = 0.0
    
    if type(x) == float or type(x) == int:
        ret_val = x
    if 'K' in x:
        if len(x) > 1:
            ret_val = float(x.replace('K', ''))
        ret_val = ret_val *1000
    if 'M' in x:
        if len(x) > 1:
            ret_val = float(x.replace('M', ''))
        ret_val = ret_val * 1000000.0
    return ret_val

Tenemos un problema con la columna "Release Clause": los valores nan son tipo float y cuando llamamos a la función value_to_float la función replace es para tipo string. Lo que podemos hacer es asignar a todos los valores nans un valor especial, quitar las letras de los valores, y por último hacer la interpolación linear.

A continuación cambiaremos los nans por la letra 'X'. **Hemos modificado ligeramente la función value_to_float()** que justo al comenzar compreba si la letra 'X' es contenida en el valor, en caso afirmativo devolvemos un 0.0 .
De esta forma se nos quedaran los nans como ceros:

In [407]:
df["Release Clause"] = df["Release Clause"].replace(np.nan, 'X')

Ahora toca convertir los valores de tipo objeto a float:

In [408]:
df["Release Clause"] = df["Release Clause"].apply(value_to_float)

Ya que la interpolación se hace sobre los valores nans, volvemos a convertir los ceros a nans:

In [409]:
df["Release Clause"] = df["Release Clause"].replace(0.0, np.nan)

Ahora ya podemos hacer la interpolación:

In [410]:
df["Release Clause"].interpolate(inplace=True)
df

Age  Overall  Potential                 Club    Value  Special  \
0       31       94         94         FC Barcelona  €110.5M     2202   
1       33       94         94             Juventus     €77M     2228   
2       26       92         93  Paris Saint-Germain  €118.5M     2143   
3       27       91         93    Manchester United     €72M     1471   
4       27       91         92      Manchester City    €102M     2281   
...    ...      ...        ...                  ...      ...      ...   
18202   19       47         65      Crewe Alexandra     €60K     1307   
18203   19       47         63       Trelleborgs FF     €60K     1098   
18204   16       47         67     Cambridge United     €60K     1189   
18205   17       47         66      Tranmere Rovers     €60K     1228   
18206   16       46         66      Tranmere Rovers     €60K     1321   

       International Reputation  Weak Foot  Skill Moves       Work Rate  \
0                           5.0        4.0          4.0  Medium/ Medium   
1                           5.0        4.0          5.0       High/ Low   
2                           5.0        5.0          5.0    High/ Medium   
3                           4.0        3.0          1.0  Medium/ Medium   
4                           4.0        5.0          4.0      High/ High   
...                         ...        ...          ...             ...   
18202                       1.0        2.0          2.0  Medium/ Medium   
18203                       1.0        2.0          2.0  Medium/ Medium   
18204                       1.0        3.0          2.0  Medium/ Medium   
18205                       1.0        3.0          2.0  Medium/ Medium   
18206                       1.0        3.0          2.0  Medium/ Medium   

      Position  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0           RF      84.0       95.0             70.0          90.0     86.0   
1           ST      84.0       94.0             89.0          81.0     87.0   
2           LW      79.0       87.0             62.0          84.0     84.0   
3           GK      17.0       13.0             21.0          50.0     13.0   
4          RCM      93.0       82.0             55.0          92.0     82.0   
...        ...       ...        ...              ...           ...      ...   
18202       CM      34.0       38.0             40.0          49.0     25.0   
18203       ST      23.0       52.0             52.0          43.0     36.0   
18204       ST      25.0       40.0             46.0          38.0     38.0   
18205       RW      44.0       50.0             39.0          42.0     40.0   
18206       CM      41.0       34.0             46.0          48.0     30.0   

       Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0           97.0   93.0        94.0         87.0         96.0          91.0   
1           88.0   81.0        76.0         77.0         94.0          89.0   
2           96.0   88.0        87.0         78.0         95.0          94.0   
3           18.0   21.0        19.0         51.0         42.0          57.0   
4           86.0   85.0        83.0         91.0         91.0          78.0   
...          ...    ...         ...          ...          ...           ...   
18202       42.0   30.0        34.0         45.0         43.0          54.0   
18203       39.0   32.0        20.0         25.0         40.0          41.0   
18204       45.0   38.0        27.0         28.0         44.0          70.0   
18205       51.0   34.0        32.0         32.0         52.0          61.0   
18206       43.0   40.0        34.0         44.0         51.0          57.0   

       SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0             86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1             91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2             90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3          

In [411]:
df.isna().sum().sum()

0

Como ya véis, nuestro data frame ya no tiene valores nans :)

---
---
---

# Casting de las columnas

Tenemos que pasar las columnas "Value" y "Wage" en valores tipo float, eliminando los signos que contienen. Para esto aplicaremos la misma función de antes:

In [412]:
df["Value"] = df["Value"].apply(value_to_float)
df

Age  Overall  Potential                 Club        Value  Special  \
0       31       94         94         FC Barcelona  110500000.0     2202   
1       33       94         94             Juventus   77000000.0     2228   
2       26       92         93  Paris Saint-Germain  118500000.0     2143   
3       27       91         93    Manchester United   72000000.0     1471   
4       27       91         92      Manchester City  102000000.0     2281   
...    ...      ...        ...                  ...          ...      ...   
18202   19       47         65      Crewe Alexandra      60000.0     1307   
18203   19       47         63       Trelleborgs FF      60000.0     1098   
18204   16       47         67     Cambridge United      60000.0     1189   
18205   17       47         66      Tranmere Rovers      60000.0     1228   
18206   16       46         66      Tranmere Rovers      60000.0     1321   

       International Reputation  Weak Foot  Skill Moves       Work Rate  \
0                           5.0        4.0          4.0  Medium/ Medium   
1                           5.0        4.0          5.0       High/ Low   
2                           5.0        5.0          5.0    High/ Medium   
3                           4.0        3.0          1.0  Medium/ Medium   
4                           4.0        5.0          4.0      High/ High   
...                         ...        ...          ...             ...   
18202                       1.0        2.0          2.0  Medium/ Medium   
18203                       1.0        2.0          2.0  Medium/ Medium   
18204                       1.0        3.0          2.0  Medium/ Medium   
18205                       1.0        3.0          2.0  Medium/ Medium   
18206                       1.0        3.0          2.0  Medium/ Medium   

      Position  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0           RF      84.0       95.0             70.0          90.0     86.0   
1           ST      84.0       94.0             89.0          81.0     87.0   
2           LW      79.0       87.0             62.0          84.0     84.0   
3           GK      17.0       13.0             21.0          50.0     13.0   
4          RCM      93.0       82.0             55.0          92.0     82.0   
...        ...       ...        ...              ...           ...      ...   
18202       CM      34.0       38.0             40.0          49.0     25.0   
18203       ST      23.0       52.0             52.0          43.0     36.0   
18204       ST      25.0       40.0             46.0          38.0     38.0   
18205       RW      44.0       50.0             39.0          42.0     40.0   
18206       CM      41.0       34.0             46.0          48.0     30.0   

       Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0           97.0   93.0        94.0         87.0         96.0          91.0   
1           88.0   81.0        76.0         77.0         94.0          89.0   
2           96.0   88.0        87.0         78.0         95.0          94.0   
3           18.0   21.0        19.0         51.0         42.0          57.0   
4           86.0   85.0        83.0         91.0         91.0          78.0   
...          ...    ...         ...          ...          ...           ...   
18202       42.0   30.0        34.0         45.0         43.0          54.0   
18203       39.0   32.0        20.0         25.0         40.0          41.0   
18204       45.0   38.0        27.0         28.0         44.0          70.0   
18205       51.0   34.0        32.0         32.0         52.0          61.0   
18206       43.0   40.0        34.0         44.0         51.0          57.0   

       SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0             86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1             91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2             90.0     96.0       94.0     

También hay que arreglar la columna "Work/Rate", la primera parte de la cual indica el ataque y la segunda parte indica la defensa. Generar tokens de estas y guardar las en dos columnas distintas, y por último podemos hace un drop de la columna original:

Generamos dos columnas a partir de los tokens, haciendo un pop de la columna original:

In [413]:
attack_defense = df.pop("Work Rate").str.split("/", n=1, expand=True)
df["Attack"] = attack_defense[0]
df["Defense"] = attack_defense[1]

Ahora toca arreglar las columnas de las que no podemos hacer casting y por lo tanto no podremos determinar valores a partir de esas características. Para arreglar este problema hacemos one-hot encoding de estas columnas.

Esas columnas precisamente son:

"Club","Attack","Defense","Position". 

Pero, ahora hay una cuestión que podemos plantear, ¿El club realmente afecta mucho o poco para determinar el valor de un jugador? A un jugador malo siempre le pagarán mal y a un jugador bueno le pagaran bien, sea cual sea el club. Eso sí, el salario puede ser ligeramente variada de un club a otro. Entonces, el club sí afecta, pero no mucho.

El problema de hacer one-hot encoding es que genera una columna para cada valor único de la columna original, puede ser que hayan miles de clubs, entonces nos generará miles columnas. Hacer one-hot encoding de la columna "Club" no sería lo correcto.

Primero eliminamos la columna "Club"

In [414]:
df=df.drop(axis=1,columns="Club")

In [415]:
col = df.pop("Position")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(col, prefix='Position').reset_index(drop=True)], axis=1, sort=False)

In [416]:
 col = df.pop("Attack")
 df = pd.concat([df.reset_index(drop=True), pd.get_dummies(col, prefix='Attack').reset_index(drop=True)], axis=1, sort=False)

In [417]:
 col = df.pop("Defense")
 df = pd.concat([df.reset_index(drop=True), pd.get_dummies(col, prefix='Defense').reset_index(drop=True)], axis=1, sort=False)

**Ya tenemos nuestros datos preparados para hacer la Regresión Lineal**
# Regresión Lineal

Antes de entrenar los datos, debemos separar esos datos para poder hacer predicciones con el test data:

In [419]:
val = df.pop("Value")
X_train, X_test, y_train, y_test = train_test_split(df, val, test_size=0.33, random_state=42)

Ahora entrenamos la regresión lineal:

In [421]:
reg = linear_model.LinearRegression().fit(X_train, y_train)

In [422]:
preds = reg.predict(X_test)

In [423]:
r2_score(preds, y_test)

0.987157008046021

Como se puede observar, el valor de acierto en hacer predicción es de unos 98,71%.